File: transpiler/javatests/com/google/j2cl/readable/java/jsfunction/JsFunctionOptimization.java
/*
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jsfunction;

import jsinterop.annotations.JsFunction;

public class JsFunctionOptimization {

  @JsFunction
  interface F {
    String m(String s);
  }

  public void main(final int r) {
    new Object() {
      void m() {
        String x = "";
        new Object() {
          void m() {
            final int var = 1;
            F f =
                new F() {

                  @Override
                  public String m(String s) {
                    final int r1 = r;
                    final int var1 = var;
                    final String x1 = x;
                    return String.valueOf(r)
                        + s
                        + x
                        + var
                        + new F() {
                          @Override
                          public String m(String s) {
                            return s + r1 + x1 + var1;
                          }
                        }.m("hello");
                  }
                };
            F f2 =
                new F() {

                  @Override
                  public String m(String s) {
                    final int r1 = r;
                    final int var1 = var;
                    final String x1 = x;
                    return String.valueOf(r)
                        + s
                        + x
                        + var
                        + new Object() {
                          public String sayHey() {
                            return "Hey";
                          }
                        }.sayHey();
                  }
                };
          }
        };
      }
    };
  }

  @JsFunction
  interface Consumer {
    <T> void accept(T t);
  }

  public void testParameterizedMethod() {
    Consumer c =
        new Consumer() {
          @Override
          public <T> void accept(T t) {}
        };
  }
}

File: transpiler/javatests/com/google/j2cl/readable/kotlin/jsfunction/JsFunctionOptimization.kt
/*
 * Copyright 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jsfunction

import jsinterop.annotations.JsFunction

@JsFunction
internal fun interface F {
  fun m(s: String): String
}

fun main(r: Int) {
  object {
    fun m() {
      val x = ""
      object {
        fun m() {
          val `var` = 1
          val f: F =
            object : F {
              override fun m(s: String): String {
                return (r.toString() +
                  s +
                  x +
                  `var` +
                  object : F {
                      override fun m(s: String): String {
                        return s + r + x + `var`
                      }
                    }
                    .m("hello"))
              }
            }
          val f2: F =
            object : F {
              override fun m(s: String): String {
                val r1 = r
                val var1 = `var`
                val x1 = x
                return (r.toString() +
                  s +
                  x +
                  `var` +
                  object : Any() {
                      fun sayHey(): String {
                        return "Hey"
                      }
                    }
                    .sayHey())
              }
            }
        }
      }
    }
  }

  // Make sure that optimized types are not leaked if they are propagated through inference.
  // TODO(b/355664391): Make sure that more complex cases of inference are handled correctly.
  val f =
    object : F {
      override fun m(s: String): String {
        return ""
      }
    }

  val array: Array<Any?> =
    arrayOf(
      object : F {
        override fun m(s: String): String {
          return ""
        }
      }
    )

  val array2 =
    arrayOf(
      object : F {
        override fun m(s: String): String {
          return ""
        }
      }
    )

  val array3 =
    arrayOf<Any?>(
      object : F {
        override fun m(s: String): String {
          return ""
        }
      }
    )

  Holder(
    object : F {
      override fun m(s: String): String {
        return ""
      }
    }
  )

  Holder(
    1,
    object : F {
      override fun m(s: String): String {
        return ""
      }
    },
  )
}

internal class Holder<T> {
  constructor(value: T)

  constructor(i: Int, vararg value: T)
}
